cardano-node/src/Cardano/Node/Handlers/TopLevel.hs

```diff
@@ -95,30 +95,40 @@ toplevelExceptionHandler prog = do
     rethrowExitCode :: ExitCode -> IO a
     rethrowExitCode = throwIO
 
-    rethrowDiffusionErrorExceptionInLinkedThreadExitSuccess :: Network.Failure -> IO a
-    rethrowDiffusionErrorExceptionInLinkedThreadExitSuccess full =
+    -- Handle errors thrown by the diffusion.
+    handleDiffusionError :: Network.Failure -> IO a
+    handleDiffusionError full =
       case full of
-        Network.DiffusionError e ->
-          case fromException (toException e) of
-            Just (ExceptionInLinkedThread _ eInner)
-              | Just exitCode <- fromException eInner
-              -> throwIO @ExitCode exitCode
-            _ -> throwIO full
-        _ -> throwIO full
+        Network.DiffusionError e
+          | Just e' <- fromException e
+          -> rethrowAsyncExceptions e'
+
+          | Just e' <- fromException e
+          -> rethrowExitCode e'
```

Comment:
I don't get these guards, aren't they the same? What are you trying to achieve here?

Reply:
They are different because the `rethrowAsyncExceptions` and `rethrowExitCode` force  `e'` to have different type (`SomeAsyncException` vs `ExitCode`).

Reply:
Perhaps `fromException` can take type arguments to make this difference explicit?